2018/10/04 - Bryan Jen, Puppet Inc.

This patch to the Ruby batch file generation is necessary when packaging
Ruby with Puppet's vanagon package builder, as used by Puppet agent,
Bolt, and PDK. When the msi installer is built, the path generated for
the installed Ruby cannot be assumed to be the same as the path in
the build environment. For that reason, we need this patch to use the
executables from the current working directory, so that it is always
using the vendored ruby installation.
---
 lib/rubygems/commands/setup_command.rb | 13 +++++++++----
 lib/rubygems/installer.rb              | 13 +++++++++----
 2 files changed, 18 insertions(+), 8 deletions(-)

diff --git a/lib/rubygems/commands/setup_command.rb b/lib/rubygems/commands/setup_command.rb
index ee6ce88eae..2cda5866de 100644
--- a/lib/rubygems/commands/setup_command.rb
+++ b/lib/rubygems/commands/setup_command.rb
@@ -256,14 +256,19 @@ def install_executables(bin_dir)
             bin_cmd_file = File.join Dir.tmpdir, "#{bin_file}.bat"
 
             File.open bin_cmd_file, 'w' do |file|
-              file.puts <<-TEXT
+              file.puts <<-SCRIPT
   @ECHO OFF
   IF NOT "%~f0" == "~f0" GOTO :WinNT
-  @"#{File.basename(Gem.ruby).chomp('"')}" "#{dest_file}" %1 %2 %3 %4 %5 %6 %7 %8 %9
+  ECHO.This version of Ruby has not been built with support for Windows 95/98/Me.
   GOTO :EOF
   :WinNT
-  @"#{File.basename(Gem.ruby).chomp('"')}" "%~dpn0" %*
-  TEXT
+  SET RUBY_EXE_PATH="%~dp0ruby.exe"
+  @%RUBY_EXE_PATH% "%~dpn0" %*
+  SCRIPT
             end
 
             install bin_cmd_file, "#{dest_file}.bat", :mode => 0755
diff --git a/lib/rubygems/installer.rb b/lib/rubygems/installer.rb
index ee5fedeb64..e0a228f9b5 100644
--- a/lib/rubygems/installer.rb
+++ b/lib/rubygems/installer.rb
@@ -754,14 +754,19 @@ def app_script_text(bin_file_name)
 
   def windows_stub_script(bindir, bin_file_name)
     ruby = Gem.ruby.gsub(/^\"|\"$/, "").tr(File::SEPARATOR, "\\")
-    return <<-TEXT
+    return <<-SCRIPT
 @ECHO OFF
 IF NOT "%~f0" == "~f0" GOTO :WinNT
-@"#{ruby}" "#{File.join(bindir, bin_file_name)}" %1 %2 %3 %4 %5 %6 %7 %8 %9
+ECHO.This version of Ruby has not been built with support for Windows 95/98/Me.
 GOTO :EOF
 :WinNT
-@"#{ruby}" "%~dpn0" %*
-TEXT
+  SET RUBY_EXE_PATH="%~dp0ruby.exe"
+@%RUBY_EXE_PATH% "%~dpn0" %*
+SCRIPT
   end
 
   ##
--
2.4.9 (Apple Git-60)
